File: src/lethe/memory/store.py
"""Main memory store coordinating all memory subsystems."""

from pathlib import Path
from typing import Optional
import lancedb
import logging

logger = logging.getLogger(__name__)

from lethe.memory.blocks import BlockManager
from lethe.memory.archival import ArchivalMemory
from lethe.memory.messages import MessageHistory
from lethe.memory.embeddings import needs_reindex, save_model_metadata


def _table_names(db: lancedb.DBConnection) -> list[str]:
    result = db.list_tables()
    return result.tables if hasattr(result, 'tables') else list(result)


class MemoryStore:
    """Unified memory store.
    
    Provides:
    - blocks: Core memory as files in workspace (identity.md, human.md, project.md, etc.)
    - archival: Long-term semantic memory with hybrid search (LanceDB)
    - messages: Conversation history (LanceDB)
    
    Blocks live in workspace for easy editing. Initialized from data/ templates.
    """
    
    def __init__(self, data_dir: str = "data/memory", workspace_dir: str = "workspace", config_dir: str = "config"):
        """Initialize memory store.
        
        Args:
            data_dir: Directory for persistent data (archival, messages)
            workspace_dir: Working directory for blocks (agent reads/writes here)
            config_dir: Directory with seed block templates
        """
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(parents=True, exist_ok=True)
        
        self.workspace_dir = Path(workspace_dir)
        self.workspace_dir.mkdir(parents=True, exist_ok=True)
        
        self.config_dir = Path(config_dir)
        
        # Connect to LanceDB (for archival and messages only)
        self.db = lancedb.connect(str(self.data_dir / "lancedb"))
        logger.info(f"Connected to LanceDB at {self.data_dir / 'lancedb'}")
        
        # Initialize blocks in workspace, copying from config/blocks/ seeds if needed
        blocks_workspace = self.workspace_dir / "memory"
        blocks_workspace.mkdir(parents=True, exist_ok=True)
        self._init_blocks_from_templates(blocks_workspace, str(self.config_dir))
        
        # Create skills and projects directories
        skills_dir = self.workspace_dir / "skills"
        skills_dir.mkdir(parents=True, exist_ok=True)
        (self.workspace_dir / "projects").mkdir(parents=True, exist_ok=True)
        self._ensure_skills_bootstrap(skills_dir)
        
        # Copy workspace seed files (questions.md, etc.) if not present
        self._init_workspace_seeds(str(self.config_dir))
        
        # Check if embedding model changed — must migrate before init
        lancedb_dir = self.data_dir / "lancedb"
        reindex = needs_reindex(lancedb_dir)
        if reindex and _table_names(self.db):
            logger.warning("Embedding model changed — migrating vector tables")
            self._migrate_tables()

        # Initialize subsystems
        self.blocks = BlockManager(blocks_workspace)
        self.archival = ArchivalMemory(self.db)
        self.messages = MessageHistory(self.db)

        if reindex:
            if self._has_note_files():
                self._reindex_notes()
            save_model_metadata(lancedb_dir)
            logger.info("Embedding model metadata saved")

        logger.info("Memory store initialized")
    
    def _init_workspace_seeds(self, config_dir: str = "config"):
        """Copy workspace seed files/directories to workspace if not present.

        Includes:
        - config/workspace/* -> workspace/*
        - config/prompts/* -> workspace/prompts/*
        """
        seeds_dir = Path(config_dir) / "workspace"
        if seeds_dir.exists():
            for seed_file in seeds_dir.rglob("*"):
                if not seed_file.is_file():
                    continue
                rel = seed_file.relative_to(seeds_dir)
                target = self.workspace_dir / rel
                if not target.exists():
                    target.parent.mkdir(parents=True, exist_ok=True)
                    target.write_text(seed_file.read_text())
                    logger.info(f"Initialized workspace file from seed: {rel}")

        prompts_dir = Path(config_dir) / "prompts"
        if prompts_dir.exists():
            for prompt_file in prompts_dir.rglob("*.md"):
                rel = prompt_file.relative_to(prompts_dir)
                target = self.workspace_dir / "prompts" / rel
                if not target.exists():
                    target.parent.mkdir(parents=True, exist_ok=True)
                    target.write_text(prompt_file.read_text())
                    logger.info(f"Initialized workspace prompt from seed: prompts/{rel}")

    def _ensure_skills_bootstrap(self, skills_dir: Path):
        """Ensure the skills directory always has a known entrypoint file."""
        readme = skills_dir / "README.md"
        if readme.exists():
            return

        readme.write_text(
            "# Skills\n\n"
            "This directory stores skill files with extended workflows and references.\n"
            "This README is intentionally always present so skills are discoverable.\n\n"
            "Use core tools to work with skills:\n"
            f"- list_directory(\"{skills_dir}/\")\n"
            f"- read_file(\"{skills_dir}/README.md\")\n"
            f"- read_file(\"{skills_dir}/<name>.md\")\n"
            f"- grep_search(\"keyword\", path=\"{skills_dir}/\")\n"
        )
        logger.info("Initialized default skills README")
    
    def _init_blocks_from_templates(self, blocks_workspace: Path, config_dir: str = "config"):
        """Copy block seeds from config/blocks/ to workspace if not present."""
        templates_dir = Path(config_dir) / "blocks"
        if not templates_dir.exists():
            logger.debug(f"No seed blocks found at {templates_dir}")
            return
        
        for template_file in templates_dir.glob("*.md"):
            target_file = blocks_workspace / template_file.name
            if not target_file.exists():
                # Copy content
                target_file.write_text(template_file.read_text())
                logger.info(f"Initialized block from seed: {template_file.name}")
                
                # Copy metadata if exists
                meta_file = template_file.with_suffix(".meta.json")
                if meta_file.exists():
                    target_meta = blocks_workspace / meta_file.name
                    target_meta.write_text(meta_file.read_text())
    
    # Blocks that rarely change — eligible for long-lived cache (1h)
    STABLE_BLOCKS = {"human"}

    @staticmethod
    def _parse_iso_timestamp(raw: str):
        """Parse ISO timestamp safely."""
        from datetime import datetime
        if not raw:
            return None
        try:
            return datetime.fromisoformat(raw.replace("Z", "+00:00"))
        except Exception:
            return None

    @staticmethod
    def _format_timestamp(dt) -> str:
        """Format timestamps with weekday (local timezone)."""
        from datetime import timezone
        if not dt:
            return ""
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        dt = dt.astimezone()  # convert to system local tz
        return dt.strftime("%a %Y-%m-%d %H:%M:%S %Z")

    def _format_block(self, block: dict) -> str:
        """Format a single block for context."""
        label = block["label"]
        value = block["value"] or ""
        description = block.get("description") or ""
        limit = block.get("limit") or 20000
        created_at = self._parse_iso_timestamp(block.get("created_at", ""))
        updated_at = self._parse_iso_timestamp(block.get("updated_at", ""))
        
        lines = [
            f"<{label}>",
            "<description>",
            description,
            "</description>",
            "<metadata>",
            f"- chars={len(value)}/{limit}",
        ]
        if created_at:
            lines.append(f"- created_at={self._format_timestamp(created_at)}")
        if updated_at:
            lines.append(f"- updated_at={self._format_timestamp(updated_at)}")
        lines.extend([
            "</metadata>",
            "<value>",
            value,
            "</value>",
            f"</{label}>",
        ])
        return "\n".join(lines)

    def get_context_for_prompt(self, max_tokens: int = 8000) -> str:
        """Get formatted memory context for LLM prompt (single string).
        
        Used by non-Anthropic models (no cache splitting needed).
        """
        stable, volatile = self.get_context_split()
        parts = [p for p in [stable, volatile] if p]
        return "\n\n".join(parts)

    def get_context_split(self) -> tuple:
        """Get memory context split into (stable, volatile) for caching.
        
        - stable: blocks that rarely change (human, tools) — 1h cache
        - volatile: blocks that change often (project, tasks) + metadata — 5m cache
        
        Returns:
            (stable_str, volatile_str) — either can be empty string
        """
        from datetime import datetime, timezone
        
        blocks = self.blocks.list_blocks()
        stable_parts = []
        volatile_parts = []
        
        if blocks:
            stable_block_strs = []
            volatile_block_strs = []
            
            for block in blocks:
                if block.get("hidden"):
                    continue
                label = block["label"]
                if label == "identity":
                    continue
                
                formatted = self._format_block(block)
                if label in self.STABLE_BLOCKS:
                    stable_block_strs.append(formatted)
                else:
                    volatile_block_strs.append(formatted)
            
            if stable_block_strs:
                stable_parts.append(
                    "<memory_blocks_stable>\n" +
                    "\n\n".join(stable_block_strs) +
                    "\n</memory_blocks_stable>"
                )
            
            if volatile_block_strs:
                volatile_parts.append(
                    "<memory_blocks>\n" +
                    "\n\n".join(volatile_block_strs) +
                    "\n</memory_blocks>"
                )
        
        # Memory metadata (volatile — counts change)
        now = datetime.now(timezone.utc)
        message_count = self.messages.count()
        archival_count = self.archival.count()
        
        last_modified = None
        for block in blocks:
            if block.get("updated_at"):
                block_time = self._parse_iso_timestamp(block["updated_at"])
                if block_time and (last_modified is None or block_time > last_modified):
                    last_modified = block_time
        if last_modified is None:
            last_modified = now
        
        metadata_lines = [
            "<memory_metadata>",
            f"- now={self._format_timestamp(now)}",
            f"- memory_blocks_last_modified={self._format_timestamp(last_modified)}",
            f"- {message_count} previous messages between you and the user are stored in recall memory (use tools to access them)",
            "- Timestamps on messages are for your reference only. Do not include timestamps in your responses.",
        ]
        if archival_count > 0:
            metadata_lines.append(f"- {archival_count} total memories you created are stored in archival memory (use tools to access them)")
        metadata_lines.append("</memory_metadata>")
        volatile_parts.append("\n".join(metadata_lines))
        
        return "\n\n".join(stable_parts), "\n\n".join(volatile_parts)
    
    def search(
        self,
        query: str,
        limit: int = 10,
        search_type: str = "hybrid"
    ) -> list[dict]:
        """Search across archival memory.
        
        Args:
            query: Search query
            limit: Max results
            search_type: "hybrid", "vector", or "fts"
            
        Returns:
            List of matching passages
        """
        return self.archival.search(query, limit=limit, search_type=search_type)
    
    def add_memory(self, text: str, metadata: Optional[dict] = None) -> str:
        """Add a memory to archival storage.
        
        Args:
            text: Memory text
            metadata: Optional metadata
            
        Returns:
            Memory ID
        """
        return self.archival.add(text, metadata=metadata)
    
    def get_recent_messages(self, limit: int = 20) -> list[dict]:
        """Get recent conversation messages.
        
        Args:
            limit: Max messages to return
            
        Returns:
            List of messages
        """
        return self.messages.get_recent(limit=limit)
    
    def add_message(self, role: str, content: str, metadata: Optional[dict] = None) -> str:
        """Add a message to history.
        
        Args:
            role: Message role (user, assistant, system, tool)
            content: Message content
            metadata: Optional metadata
            
        Returns:
            Message ID
        """
        return self.messages.add(role, content, metadata=metadata)

    def _migrate_tables(self):
        """Re-embed archival and message tables with the new embedding model.

        Exports text data, drops tables (dimension changed), then re-inserts
        with fresh vectors. Notes are file-backed and reindexed separately.
        """
        from lethe.memory.embeddings import embed, EMBEDDING_DIM

        for table_name in list(_table_names(self.db)):
            try:
                table = self.db.open_table(table_name)
                arrow = table.to_arrow()
                columns = arrow.column_names
            except Exception as e:
                logger.warning(f"Could not read table {table_name}, dropping: {e}")
                self.db.drop_table(table_name)
                continue

            if "vector" not in columns:
                continue

            ids = arrow.column("id").to_pylist()
            rows = []
            for i in range(arrow.num_rows):
                if ids[i] == "_init_":
                    continue
                row = {col: arrow.column(col)[i].as_py() for col in columns if col != "vector"}
                rows.append(row)

            self.db.drop_table(table_name)
            logger.info(f"Migrating {table_name}: {len(rows)} rows")

            if not rows:
                continue

            text_col = "text" if "text" in columns else "content"
            for row in rows:
                row["vector"] = embed(row.get(text_col, "") or "", is_query=False)

            init_row = {col: "" for col in columns if col != "vector"}
            init_row["id"] = "_init_"
            init_row["vector"] = [0.0] * EMBEDDING_DIM
            self.db.create_table(table_name, data=[init_row], exist_ok=True)
            table = self.db.open_table(table_name)
            table.add(rows)

            if text_col in columns:
                try:
                    table.create_fts_index(text_col, replace=True)
                except Exception:
                    pass

            logger.info(f"Migrated {table_name}: {len(rows)} rows re-embedded")

    def _has_note_files(self) -> bool:
        from lethe.paths import notes_dir
        nd = notes_dir()
        return nd.exists() and any(nd.rglob("*.md"))

    def _reindex_notes(self):
        try:
            from lethe.memory.notes import NoteStore
            notes = NoteStore(db=self.db)
            count = notes.reindex()
            logger.info(f"Reindexed {count} notes")
        except Exception as e:
            logger.warning(f"Note reindex failed: {e}")

File: src/store.rs
use std::fs;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Local, Utc};
use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::config::Settings;
use crate::memory::archival::{ArchivalEntry, ArchivalError, ArchivalMemory};
use crate::memory::messages::{MessageHistory, MessageHistoryError, StoredMessage};
use crate::memory::notes::{NoteError, NoteSearchResult, NoteStore};
use crate::memory::{BlockManager, MemoryBlock, MemoryDb, MemoryError};

#[derive(Debug, Error)]
pub enum MemoryStoreError {
    #[error(transparent)]
    Blocks(#[from] MemoryError),
    #[error(transparent)]
    Archival(#[from] ArchivalError),
    #[error(transparent)]
    Messages(#[from] MessageHistoryError),
    #[error(transparent)]
    Notes(#[from] NoteError),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Sqlite(#[from] rusqlite::Error),
}

pub type MemoryStoreResult<T> = Result<T, MemoryStoreError>;

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct MemoryStats {
    pub memory_blocks: usize,
    pub archival_memories: usize,
    pub message_history: usize,
    pub total_messages: usize,
    pub notes: usize,
}

#[derive(Debug)]
pub struct MemoryStore {
    pub blocks: BlockManager,
    pub archival: ArchivalMemory,
    pub messages: MessageHistory,
    pub notes: NoteStore,
    db_path: PathBuf,
    memory_data_path: PathBuf,
    workspace_dir: PathBuf,
}

impl MemoryStore {
    pub fn from_settings(settings: &Settings) -> MemoryStoreResult<Self> {
        Self::open_with_data_path(
            &settings.workspace_dir,
            &settings.db_path,
            &settings.notes_dir,
            settings.memory_dir.join("lethe-memory.db"),
        )
    }

    pub fn open(
        workspace_dir: impl Into<PathBuf>,
        db_path: impl Into<PathBuf>,
        notes_dir: impl Into<PathBuf>,
    ) -> MemoryStoreResult<Self> {
        let workspace_dir = workspace_dir.into();
        let db_path = db_path.into();
        let notes_dir = notes_dir.into();
        let memory_data_path = memory_data_path_for(&db_path);
        Self::open_with_data_path(workspace_dir, db_path, notes_dir, memory_data_path)
    }

    pub fn open_with_data_path(
        workspace_dir: impl Into<PathBuf>,
        db_path: impl Into<PathBuf>,
        notes_dir: impl Into<PathBuf>,
        memory_data_path: impl Into<PathBuf>,
    ) -> MemoryStoreResult<Self> {
        let workspace_dir = workspace_dir.into();
        let db_path = db_path.into();
        let notes_dir = notes_dir.into();
        let memory_data_path = memory_data_path.into();

        fs::create_dir_all(&workspace_dir)?;
        fs::create_dir_all(workspace_dir.join("projects"))?;
        ensure_skills_bootstrap(&workspace_dir.join("skills"))?;

        let blocks = BlockManager::new(workspace_dir.join("memory"))?;
        blocks.init_embedded_defaults()?;
        let memory_db = MemoryDb::open(&memory_data_path)?;
        let archival = ArchivalMemory::from_db(memory_db.clone());
        let messages = MessageHistory::open(&memory_data_path)?;
        let notes = NoteStore::new_with_db(notes_dir, memory_db)?;

        Ok(Self {
            blocks,
            archival,
            messages,
            notes,
            db_path,
            memory_data_path,
            workspace_dir,
        })
    }

    pub fn db_path(&self) -> &Path {
        &self.db_path
    }

    pub fn memory_data_path(&self) -> &Path {
        &self.memory_data_path
    }

    pub fn workspace_dir(&self) -> &Path {
        &self.workspace_dir
    }

    pub fn stats(&self) -> MemoryStoreResult<MemoryStats> {
        let memory_blocks = self.blocks.list_blocks(true)?.len();
        let archival_memories = self.archival.count()?;
        let message_history = self.messages.count()?;
        let notes = self.notes.list_notes(None)?.len();
        Ok(MemoryStats {
            memory_blocks,
            archival_memories,
            message_history,
            total_messages: message_history,
            notes,
        })
    }

    pub fn get_context_for_prompt(&self) -> MemoryStoreResult<String> {
        let (stable, volatile) = self.get_context_split()?;
        Ok([stable, volatile]
            .into_iter()
            .filter(|part| !part.trim().is_empty())
            .collect::<Vec<_>>()
            .join("\n\n"))
    }

    pub fn get_context_split(&self) -> MemoryStoreResult<(String, String)> {
        let blocks = self.blocks.list_blocks(false)?;
        let mut stable_blocks = Vec::new();
        let mut volatile_blocks = Vec::new();

        for block in &blocks {
            if block.hidden || block.label == "identity" {
                continue;
            }
            let formatted = format_block(block);
            if is_stable_block(&block.label) {
                stable_blocks.push(formatted);
            } else {
                volatile_blocks.push(formatted);
            }
        }

        let mut stable_parts = Vec::new();
        let mut volatile_parts = Vec::new();
        if !stable_blocks.is_empty() {
            stable_parts.push(format!(
                "<memory_blocks_stable>\n{}\n</memory_blocks_stable>",
                stable_blocks.join("\n\n")
            ));
        }
        if !volatile_blocks.is_empty() {
            volatile_parts.push(format!(
                "<memory_blocks>\n{}\n</memory_blocks>",
                volatile_blocks.join("\n\n")
            ));
        }

        volatile_parts.push(self.format_memory_metadata(&blocks)?);

        Ok((stable_parts.join("\n\n"), volatile_parts.join("\n\n")))
    }

    pub fn search_archival(
        &self,
        query: &str,
        limit: usize,
        tags: Option<&[String]>,
    ) -> MemoryStoreResult<Vec<ArchivalEntry>> {
        self.archival
            .search(query, normalized_limit(limit, 10), tags)
            .map_err(Into::into)
    }

    pub fn search_messages(
        &self,
        query: &str,
        limit: usize,
        role: Option<&str>,
    ) -> MemoryStoreResult<Vec<StoredMessage>> {
        self.messages
            .search(query, normalized_limit(limit, 20), role)
            .map_err(Into::into)
    }

    pub fn search_notes(
        &self,
        query: &str,
        tags: Option<&[String]>,
        limit: usize,
    ) -> MemoryStoreResult<Vec<NoteSearchResult>> {
        self.notes
            .search(query, tags, normalized_limit(limit, 5))
            .map_err(Into::into)
    }

    fn format_memory_metadata(&self, blocks: &[MemoryBlock]) -> MemoryStoreResult<String> {
        let now = Utc::now();
        let last_modified = blocks
            .iter()
            .filter_map(|block| block.updated_at.or(block.created_at))
            .max()
            .unwrap_or(now);
        let message_count = self.messages.count()?;
        let archival_count = self.archival.count()?;

        let mut lines = vec![
            "<memory_metadata>".to_string(),
            format!("- now={}", format_timestamp(now)),
            format!(
                "- memory_blocks_last_modified={}",
                format_timestamp(last_modified)
            ),
            format!(
                "- {message_count} previous messages between you and the user are stored in recall memory (use tools to access them)"
            ),
            "- Timestamps on messages are for your reference only. Do not include timestamps in your responses.".to_string(),
        ];
        if archival_count > 0 {
            lines.push(format!(
                "- {archival_count} total memories you created are stored in archival memory (use tools to access them)"
            ));
        }
        lines.push("</memory_metadata>".to_string());
        Ok(lines.join("\n"))
    }
}

fn ensure_skills_bootstrap(skills_dir: &Path) -> std::io::Result<()> {
    fs::create_dir_all(skills_dir)?;
    let readme = skills_dir.join("README.md");
    if readme.exists() {
        return Ok(());
    }
    fs::write(
        readme,
        "# Skills\n\n\
This directory stores skill files with extended workflows and references.\n\
This README is intentionally always present so skills are discoverable.\n\n\
Use core tools to work with skills:\n\
- list_directory(\"workspace/skills/\")\n\
- read_file(\"workspace/skills/README.md\")\n\
- read_file(\"workspace/skills/<name>.md\")\n\
- grep_search(\"keyword\", path=\"workspace/skills/\")\n",
    )
}

fn format_block(block: &MemoryBlock) -> String {
    let mut lines = vec![
        format!("<{}>", block.label),
        "<description>".to_string(),
        block.description.clone(),
        "</description>".to_string(),
        "<metadata>".to_string(),
        format!("- chars={}/{}", block.value.len(), block.limit),
    ];
    if let Some(created_at) = block.created_at {
        lines.push(format!("- created_at={}", format_timestamp(created_at)));
    }
    if let Some(updated_at) = block.updated_at {
        lines.push(format!("- updated_at={}", format_timestamp(updated_at)));
    }
    lines.extend([
        "</metadata>".to_string(),
        "<value>".to_string(),
        block.value.clone(),
        "</value>".to_string(),
        format!("</{}>", block.label),
    ]);
    lines.join("\n")
}

fn is_stable_block(label: &str) -> bool {
    label == "human"
}

fn format_timestamp(time: DateTime<Utc>) -> String {
    time.with_timezone(&Local)
        .format("%a %Y-%m-%d %H:%M:%S %Z")
        .to_string()
}

fn memory_data_path_for(db_path: &Path) -> PathBuf {
    db_path
        .parent()
        .map(|parent| parent.join("memory").join("lethe-memory.db"))
        .unwrap_or_else(|| PathBuf::from("memory").join("lethe-memory.db"))
}

fn normalized_limit(limit: usize, default: usize) -> usize {
    if limit == 0 { default } else { limit }
}

#[cfg(test)]
mod tests {
    use serde_json::json;
    use tempfile::tempdir;

    use super::*;

    fn store() -> (tempfile::TempDir, MemoryStore) {
        let tmp = tempdir().unwrap();
        let workspace = tmp.path().join("workspace");
        let db_path = tmp.path().join("data").join("lethe.db");
        let notes_dir = workspace.join("notes");
        let store = MemoryStore::open(workspace, db_path, notes_dir).unwrap();
        (tmp, store)
    }

    #[test]
    fn open_initializes_workspace_blocks_and_stats() {
        let (tmp, store) = store();

        assert!(store.workspace_dir().join("skills/README.md").exists());
        assert!(store.workspace_dir().join("projects").exists());
        let labels = store
            .blocks
            .list_blocks(true)
            .unwrap()
            .into_iter()
            .map(|block| block.label)
            .collect::<Vec<_>>();
        assert_eq!(labels, vec!["human", "identity", "project"]);

        store
            .notes
            .create("Test note", "body", &["skill".to_string()], None)
            .unwrap();
        let stats = store.stats().unwrap();
        assert_eq!(stats.memory_blocks, 3);
        assert_eq!(stats.notes, 1);
        assert_eq!(stats.total_messages, 0);
        drop(tmp);
    }

    #[test]
    fn context_split_formats_blocks_and_metadata_counts() {
        let (_tmp, store) = store();
        store
            .blocks
            .update("human", Some("User prefers concise answers."), None)
            .unwrap();
        store
            .blocks
            .update("project", Some("Port Lethe to Rust."), None)
            .unwrap();
        store.messages.add("user", "hello", None).unwrap();
        store
            .archival
            .add(
                "Remember the Rust port context.",
                Some(json!({"source": "test"})),
                &[],
            )
            .unwrap();

        let (stable, volatile) = store.get_context_split().unwrap();
        assert!(stable.contains("<memory_blocks_stable>"));
        assert!(stable.contains("User prefers concise answers."));
        assert!(volatile.contains("<memory_blocks>"));
        assert!(volatile.contains("Port Lethe to Rust."));
        assert!(volatile.contains("- 1 previous messages"));
        assert!(volatile.contains("- 1 total memories"));
        assert!(!stable.contains("<identity>"));

        let combined = store.get_context_for_prompt().unwrap();
        assert!(combined.contains("<memory_metadata>"));
    }
}
